### PR TITLE
Add min and max cohort size options to haplotypes, and add support in h12, h1x and fst functions 

### LINKED ISSUE #370: Add min_cohort_size and max_cohort_size to h12 and fst gwss functions

### PR DESCRIPTION
**Pull request by sanjaynagi**
- Ive moved the min and max cohort size from the iHS function to the `haplotypes()` function as in `snp_calls()`, and added support for in H12 and H1X. 

PS. the doc generator is so much better!

Resolves #370.

### CODE REVIEW

**Comment by sanjaynagi** [comment]
> PS. the doc generator is so much better!

That is, until the linter gets me! :P

File: malariagen_data/anopheles.py [hunk 1]
```diff
@@ -7236,8 +7290,8 @@ def ihs_gwss(
         gap_scale: ihs_params.gap_scale = ihs_params.gap_scale_default,
         include_edges: ihs_params.include_edges = True,
         use_threads: ihs_params.use_threads = True,
-        min_cohort_size: base_params.min_cohort_size = 15,
-        max_cohort_size: base_params.max_cohort_size = 50,
+        min_cohort_size: Optional[base_params.min_cohort_size] = 15,
+        max_cohort_size: Optional[base_params.max_cohort_size] = 50,
```
Comment (alimanfoo) [inline malariagen_data/anopheles.py:7294]: Suggest defining these defaults on the `ihs_params` class, i.e., `ihs_params.min_cohort_size_default` and `ihs_params.max_cohort_size_default`.

File: malariagen_data/anopheles.py [hunk 2]
```diff
@@ -7406,8 +7446,8 @@ def plot_ihs_gwss_track(
...
```
Comment (alimanfoo) [inline malariagen_data/anopheles.py:7450]: Suggest defining defaults on `ihs_params` as above.

File: malariagen_data/anopheles.py [hunk 3]
```diff
@@ -7522,7 +7562,7 @@ def plot_ihs_gwss(
         gap_scale: ihs_params.gap_scale = ihs_params.gap_scale_default,
         include_edges: ihs_params.include_edges = True,
         use_threads: ihs_params.use_threads = True,
-        min_cohort_size: base_params.min_cohort_size = 15,
+        min_cohort_size: Optional[base_params.min_cohort_size] = 15,
         max_cohort_size: base_params.max_cohort_size = 50,
```
Comment (alimanfoo) [inline malariagen_data/anopheles.py:7566]: Suggest defining defaults on `ihs_params` as above.

Also, `max_cohort_size` type should be `Optional[base_params.max_cohort_size]` here?

**Review by alimanfoo — COMMENTED** [review]
Awesome, thanks Sanjinho :soccer: 

While you're here, could you give the fst_gwss functions the same treatment?

**Comment by alimanfoo** [comment]
Hi @sanjaynagi, just a heads-up I've updated this PR branch from master, which includes changes to the poetry environment, so please `git pull` and `poetry install` before any further commits.

**Comment by sanjaynagi** [comment]
Cool, have updated the local branch and made the suggestions :)

**Comment by sanjaynagi** [comment]
Forgot to do Fst, done now!

File: malariagen_data/anopheles.py [hunk 4]
```diff
@@ -248,6 +248,31 @@ class h12_params:
...
         """,
     ]
+    cohort_size: TypeAlias = Annotated[
+        Optional[int],
+        """
+        Randomly down-sample to this value if the number of samples in the
+        cohort is greater. Raise an error if the number of samples is less
+        than this value.
+        """,
+    ]
```
Comment (alimanfoo) [inline malariagen_data/anopheles.py:258]:
```suggestion
```

This parameter is already defined on the `base_params` class, don't think we need to duplicate here. Can just add the default values here.

File: malariagen_data/anopheles.py [hunk 5]
```diff
@@ -656,6 +705,22 @@ class ihs_params:
...
+        less than this value.
+        """,
+    ]
+    max_cohort_size: TypeAlias = Annotated[
+        int,
+        """
+        Randomly down-sample to this value if the number of samples in the
+        cohort is greater.
+        """,
+    ]
```
Comment (alimanfoo) [inline malariagen_data/anopheles.py:721]:
```suggestion
```

As above, can remove these.

File: malariagen_data/anopheles.py [hunk 6]
```diff
@@ -5348,7 +5413,13 @@ def fst_gwss(
...
         sample_sets: Optional[base_params.sample_sets] = None,
         site_mask: base_params.site_mask = DEFAULT,
-        cohort_size: base_params.cohort_size = 30,
+        cohort_size: Optional[fst_params.cohort_size] = fst_params.cohort_size_default,
+        min_cohort_size: Optional[
+            fst_params.min_cohort_size
+        ] = fst_params.min_cohort_size_default,
+        max_cohort_size: Optional[
+            fst_params.max_cohort_size
+        ] = fst_params.max_cohort_size_default,
```
Comment (alimanfoo) [inline malariagen_data/anopheles.py:5422]:
```suggestion
        cohort_size: Optional[base_params.cohort_size] = fst_params.cohort_size_default,
        min_cohort_size: Optional[
            base_params.min_cohort_size
        ] = fst_params.min_cohort_size_default,
        max_cohort_size: Optional[
            base_params.max_cohort_size
        ] = fst_params.max_cohort_size_default,
```

File: malariagen_data/anopheles.py [hunk 7]
```diff
@@ -6622,7 +6733,13 @@ def h12_calibration(
...
         sample_query: Optional[base_params.sample_query] = None,
         sample_sets: Optional[base_params.sample_sets] = None,
-        cohort_size: base_params.cohort_size = 30,
+        cohort_size: Optional[h12_params.cohort_size] = h12_params.cohort_size_default,
+        min_cohort_size: Optional[
+            h12_params.min_cohort_size
+        ] = h12_params.min_cohort_size_default,
+        max_cohort_size: Optional[
+            h12_params.max_cohort_size
+        ] = h12_params.max_cohort_size_default,
```
Comment (alimanfoo) [inline malariagen_data/anopheles.py:6742]:
```suggestion
        cohort_size: Optional[base_params.cohort_size] = h12_params.cohort_size_default,
        min_cohort_size: Optional[
            base_params.min_cohort_size
        ] = h12_params.min_cohort_size_default,
        max_cohort_size: Optional[
            base_params.max_cohort_size
        ] = h12_params.max_cohort_size_default,
```

**Comment by alimanfoo** [comment]
Tests failing:

```
=========================== short test summary info ============================
FAILED tests/test_af1.py::test_h1x_gwss - ValueError: not enough samples (1) for minimum cohort size (15)
FAILED tests/test_af1.py::test_fst_gwss - ValueError: not enough samples (1) for minimum cohort size (15)
=========== 2 failed, 721 passed, 684 warnings in 3484.94s (0:58:04) ===========
```

**Comment by sanjaynagi** [comment]
> Thanks Sanj. Can avoid some duplication here as the parameter definitions (types and documentation) are all in common and  so we can use base_params. So just the default values need to go on the analysis-specific params classes.

Apologies, my vscode was telling me those variables were undefined when setting the defaults, so I got a bit concerned and decided to add them to each param class. Glad it doesn't work that way. 

Tests should be fixed, although I dont know why there was only 1 sample coming through, assume thats normal for that cohort.
Thanks!

**Comment by sanjaynagi** [comment]
oh yeah, @alimanfoo , flake8 is failing as theyre now coming up as undefined variables.....

**Comment by alimanfoo** [comment]
Need to do like...

```
min_cohort_size_default: base_params.min_cohort_size = 15
```

**Comment by alimanfoo** [comment]
> Need to do like...
> 
> ```
> min_cohort_size_default: base_params.min_cohort_size = 15
> ```

Sorry, missed that in the suggestions I made.

**Comment by sanjaynagi** [comment]
> Sorry, missed that in the suggestions I made.

No worries, pushed changes now :)

**Review by alimanfoo — APPROVED** [review]
Tidy, thanks Sanj.

**Comment by alimanfoo** [comment]
Will merge if CI passes.

**Pull request merged**